File: go/pkg/schema/structcounttree.go
package schema

type structCountTree struct {
	structName   string // for debugging, delete.
	fieldCount   uint
	structFields []structCountTree
}

func schemaToStructCountTree(src *FieldType, dst *structCountTree, stack *recurseStack) {
	switch {
	case src.Primitive != nil: // nothing to do

	case src.StructDef != nil:
		if stack.asMap[src.StructDef.Name] {
			return
		}

		dst.structName = src.StructDef.Name
		dst.fieldCount = uint(len(src.StructDef.Fields))
		//dst.structFields = make([]structCountTree, len(src.StructDef.Fields))

		stack.asStack = append(stack.asStack, src.StructDef.Name)
		stack.asMap[src.StructDef.Name] = true

		for _, field := range src.StructDef.Fields {
			subDst := structCountTree{}
			schemaToStructCountTree(&field.FieldType, &subDst, stack)
			if subDst.fieldCount != 0 {
				dst.structFields = append(dst.structFields, subDst)
			}
		}

		stack.asStack = stack.asStack[:len(stack.asStack)-1]
		delete(stack.asMap, src.StructDef.Name)

	case src.Array != nil:
		schemaToStructCountTree(&src.Array.ElemType, dst, stack)

	case src.MultimapDef != nil:
		if stack.asMap[src.MultimapDef.Name] {
			return
		}

		stack.asStack = append(stack.asStack, src.MultimapDef.Name)
		stack.asMap[src.MultimapDef.Name] = true

		schemaToStructCountTree(&src.MultimapDef.Key.Type, dst, stack)
		schemaToStructCountTree(&src.MultimapDef.Value.Type, dst, stack)

		stack.asStack = stack.asStack[:len(stack.asStack)-1]
		delete(stack.asMap, src.MultimapDef.Name)

	default:
		panic("unknown FieldType")
	}
}

File: rust/stef-core/src/schema/structcounttree.rs
use crate::schema::FieldType;

/// Intermediate DFS tree used to linearize wire schema struct counts.
#[derive(Debug, Clone, Default)]
pub(crate) struct StructCountTree {
    pub struct_name: String,
    pub field_count: u32,
    pub struct_fields: Vec<StructCountTree>,
}

pub(crate) fn schema_to_struct_count_tree(
    src: &FieldType,
    dst: &mut Vec<StructCountTree>,
    as_map: &mut std::collections::HashSet<String>,
) {
    if src.primitive.is_some() {
        return;
    }
    if !src.r#struct.is_empty() {
        if as_map.contains(&src.r#struct) {
            return;
        }
        as_map.insert(src.r#struct.clone());
        dst.push(StructCountTree { struct_name: src.r#struct.clone(), field_count: 0, struct_fields: Vec::new() });
        return;
    }
    if let Some(arr) = &src.array {
        schema_to_struct_count_tree(&arr.elem_type, dst, as_map);
        return;
    }
    if !src.multimap.is_empty() {
        if as_map.contains(&src.multimap) {
            return;
        }
        as_map.insert(src.multimap.clone());
    }
}
